### PR TITLE
Display `extra-words` in `detection_log` if present

### LINKED ISSUE #4400: `extra-words` does not show up in `detection_log` properly
### Description

> Please leave a brief description of the bug or feature request:

The `detection-log` attribute is not populated properly in `extra-words` case and a few other cases.
We still:

-  detect this is a `extra-word` correctly with the `--todo` option
- ``--license-text-diagnostics`` shows us where the extra word is (see `Neither the name of [filesize] nor the names of its\r\n  contributors`) here `filesize` is the extra word
- score is reduced correctly even though `matched_coverage` is perfect.

### How To Reproduce

> Tell us how to reproduce the issue. 

Package to scan: https://registry.npmjs.org/filesize/-/filesize-8.0.7.tgz

scancode options: `scancode -clpieu --license-text --license-diagnostics --license-text-diagnostics --license-references --todo --summary --classify -n 8 /path-to/filesize-8.0.7.tgz-extract/ --json-pp filesize-8.0.7.scan-results.json`

Example file level license detection for the above package with extra words:

```
     "path": "filesize-8.0.7.tgz-extract/package/LICENSE",
      "detected_license_expression": "bsd-new",
      "detected_license_expression_spdx": "BSD-3-Clause",
      "license_detections": [
        {
          "license_expression": "bsd-new",
          "license_expression_spdx": "BSD-3-Clause",
          "matches": [
            {
              "license_expression": "bsd-new",
              "license_expression_spdx": "BSD-3-Clause",
              "from_file": "filesize-8.0.7.tgz-extract/package/LICENSE",
              "start_line": 4,
              "end_line": 27,
              "matcher": "2-aho",
              "score": 99.53,
              "matched_length": 210,
              "match_coverage": 100.0,
              "rule_relevance": 100,
              "rule_identifier": "bsd-new_158.RULE",
              "rule_url": "https://github.com/nexB/scancode-toolkit/tree/develop/src/licensedcode/data/rules/bsd-new_158.RULE",
              "matched_text": "Redistribution and use in source and binary forms, with or without\r\nmodification, are permitted provided that the following conditions are met:\r\n\r\n* Redistributions of source code must retain the above copyright notice, this\r\n  list of conditions and the following disclaimer.\r\n\r\n* Redistributions in binary form must reproduce the above copyright notice,\r\n  this list of conditions and the following disclaimer in the documentation\r\n  and/or other materials provided with the distribution.\r\n\r\n* Neither the name of filesize nor the names of its\r\n  contributors may be used to endorse or promote products derived from\r\n  this software without specific prior written permission.\r\n\r\nTHIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS \"AS IS\"\r\nAND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE\r\nIMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE\r\nDISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE\r\nFOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL\r\nDAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR\r\nSERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER\r\nCAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,\r\nOR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE\r\nOF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.",
              "matched_text_diagnostics": "Redistribution and use in source and binary forms, with or without\r\nmodification, are permitted provided that the following conditions are met:\r\n\r\n* Redistributions of source code must retain the above copyright notice, this\r\n  list of conditions and the following disclaimer.\r\n\r\n* Redistributions in binary form must reproduce the above copyright notice,\r\n  this list of conditions and the following disclaimer in the documentation\r\n  and/or other materials provided with the distribution.\r\n\r\n* Neither the name of [filesize] nor the names of its\r\n  contributors may be used to endorse or promote products derived from\r\n  this software without specific prior written permission.\r\n\r\nTHIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS \"AS IS\"\r\nAND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE\r\nIMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE\r\nDISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE\r\nFOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL\r\nDAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR\r\nSERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER\r\nCAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,\r\nOR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE\r\nOF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE."
            }
          ],
          "detection_log": [],
          "identifier": "bsd_new-fbfc5955-0c63-4c98-2ce9-08e1e1796f50"
        }
      ],
      "for_todo": [
        "bsd_new-fbfc5955-0c63-4c98-2ce9-08e1e1796f50"
      ],
```

``--todo`` output for the same:

```
"todo": [
    {
      "detection_id": "bsd_new-fbfc5955-0c63-4c98-2ce9-08e1e1796f50",
      "review_comments": {
        "extra-words": "The license detection is conclusive with high confidence because all the rule text is matched, but some unknown extra words have been inserted in the text, which needs to be reviewed. "
      },
      "detection": {
        "license_expression": "bsd-new",
        "license_expression_spdx": "BSD-3-Clause",
        "matches": [
          {
            "license_expression": "bsd-new",
            "license_expression_spdx": "BSD-3-Clause",
            "from_file": "filesize-8.0.7.tgz-extract/package/LICENSE",
            "start_line": 4,
            "end_line": 27,
            "is_license_text": true,
            "matcher": "2-aho",
            "score": 99.53,
            "matched_length": 210,
            "rule_length": 210,
            "match_coverage": 100.0,
            "rule_relevance": 100,
            "rule_identifier": "bsd-new_158.RULE",
            "rule_url": "https://github.com/nexB/scancode-toolkit/tree/develop/src/licensedcode/data/rules/bsd-new_158.RULE",
            "rule_notes": null,
            "referenced_filenames": [],
            "matched_text": "Redistribution and use in source and binary forms, with or without\r\nmodification, are permitted provided that the following conditions are met:\r\n\r\n* Redistributions of source code must retain the above copyright notice, this\r\n  list of conditions and the following disclaimer.\r\n\r\n* Redistributions in binary form must reproduce the above copyright notice,\r\n  this list of conditions and the following disclaimer in the documentation\r\n  and/or other materials provided with the distribution.\r\n\r\n* Neither the name of filesize nor the names of its\r\n  contributors may be used to endorse or promote products derived from\r\n  this software without specific prior written permission.\r\n\r\nTHIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS \"AS IS\"\r\nAND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE\r\nIMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE\r\nDISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE\r\nFOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL\r\nDAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR\r\nSERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER\r\nCAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,\r\nOR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE\r\nOF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.",
            "rule_text": "Redistribution and use in source and binary forms, with or without\nmodification, are permitted provided that the following conditions\nare met:\n\nRedistributions of source code must retain the above copyright\nnotice, this list of conditions and the following disclaimer.\n\nRedistributions in binary form must reproduce the above copyright\nnotice, this list of conditions and the following disclaimer in the\ndocumentation and/or other materials provided with the distribution.\n\nNeither the name of nor the names of its\ncontributors may be used to endorse or promote products derived from\nthis software without specific prior written permission.\n\nTHIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS \"AS IS\"\nAND ANY EXPRESS OR IMPLIED WARRANTIES,\nINCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY\nAND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL\nTHE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,\nINCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,\nBUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS\nOF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND\nON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR\nTORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE\nUSE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE."
          }
        ],
        "detection_log": [],
        "identifier": "bsd_new-fbfc5955-0c63-4c98-2ce9-08e1e1796f50"
      }
    }
  ],
```

Here the expected behaviour would be `"detection_log": ["extra-words"],` in all cases.

### PR DESCRIPTION
**Pull request by alok1304**
- I added `extra_words` in  secondary types of Detections.
- I remove `2-aho` from matcher because previously, due to this, we get `perfect-detection` because of this https://github.com/aboutcode-org/scancode-toolkit/blob/develop/src/licensedcode/detection.py#L1729 , here `2-aho` matcher  included in this. 
update: when I do this many test cases are failing.
these work fine when our matcher is `3-seq`.

Fixes #4400

### Tasks

* [x] Reviewed [contribution guidelines](https://github.com/nexB/scancode-toolkit/blob/develop/CONTRIBUTING.rst)
* [x] PR is descriptively titled 📑 and links the original issue above 🔗
* [ ] Tests pass -- look for a green checkbox ✔️ a few minutes after opening your PR
  Run [tests](https://scancode-toolkit.readthedocs.io/en/latest/contribute/contrib_dev.html#running-tests) locally to check for errors. 
* [x] Commits are in uniquely-named feature branch and has no merge conflicts 📁
* [ ] Updated documentation pages (if applicable)
* [ ] Updated CHANGELOG.rst (if applicable)

Signed-off-by: Alok Kumar [alokkumarjipura9973@gmail.com](mailto:alokkumarjipura9973@gmail.com)

### CODE REVIEW

File: src/licensedcode/detection.py [hunk 1]
```diff
@@ -1725,6 +1732,10 @@ def analyze_detection(license_matches, package_license=False):
         license_matches=license_matches
     ):
         return DetectionCategory.LICENSE_CLUES.value
+
```
Comment (AyanSinhaMahapatra) [inline src/licensedcode/detection.py]: This ordering is wrong, if there are license matches with unknown matches, low quality matches they always take precedence over a `extra-words` case. Please revert back this change.

See also test failures being caused for this change: https://dev.azure.com/nexB/scancode-toolkit/_build/results?buildId=16270&view=logs&j=1412dbfa-421a-5a40-8ddf-7073fc46aa19

File: src/licensedcode/detection.py [hunk 2]
```diff
@@ -1746,10 +1757,6 @@ def analyze_detection(license_matches, package_license=False):
     ):
         return DetectionCategory.IMPERFECT_COVERAGE.value
 
-    # Case where at least one of the match have extra words
```
Comment (AyanSinhaMahapatra) [inline src/licensedcode/detection.py]: This is the correct order for `extra-words`, please move this back here.

**Review by AyanSinhaMahapatra — CHANGES_REQUESTED** [review]
@alok1304 thanks! See comments for changes.

Please always run a subset of the tests which could be effected locally and inspect failures before raising a PR, and make sure you also regenerate test expectations for these failures, and verify that they are intended. See https://scancode-toolkit.readthedocs.io/en/stable/contribute/contrib_dev.html#running-tests 

Please also add a test from https://github.com/aboutcode-org/scancode-toolkit/issues/4400 just focused on extra words, without tests it's much more work for reviewers to verify this is working correctly.

File: src/licensedcode/detection.py (hunk 2 above)
Comment (alok1304) [reply src/licensedcode/detection.py]: When I keep this at this place then I didn't get extra_words in detection_log. when I check this analyze_detection function they return `perfect-detection` , that why we not received `extra-words` in detection_log.

Comment (alok1304) [reply src/licensedcode/detection.py]: This is due to this https://github.com/aboutcode-org/scancode-toolkit/blob/develop/src/licensedcode/detection.py#L1730 
they return `perfect-detection` as you given the file they contain `matcher: 2-aho` but still this condition is executed
and they return `perfect-detection` at this point so they are unable move on condition like they have extra-words condition https://github.com/aboutcode-org/scancode-toolkit/blob/develop/src/licensedcode/detection.py#L1749 .

File: src/licensedcode/detection.py (hunk 1 above)
Comment (alok1304) [reply src/licensedcode/detection.py]: yes, you are right

File: src/licensedcode/detection.py (hunk 2 above)
Comment (alok1304) [reply src/licensedcode/detection.py]: I got this, this is happening due to this https://github.com/aboutcode-org/scancode-toolkit/blob/develop/src/licensedcode/detection.py#L1074 , because we include this `2-aho` as `perfect-detection`  that's why we receiving `perfect-detection`.

**Comment by alok1304** [comment]
still, we are not getting extra-words in `detection_log` because of this
```
 "matcher": "2-aho",
 ```
 
 I am trying to find examples files where extra-words are present.

**Comment by alok1304** [comment]
@AyanSinhaMahapatra I added test when we get `extra-words` and one more things, this happens only when matcher is `3-seq`.
```
{
      "identifier": "bsd_new-95249a8d-f533-e7c7-159a-9b6e173cba42",
      "license_expression": "bsd-new",
      "license_expression_spdx": "BSD-3-Clause",
      "detection_count": 3,
      "detection_log": [
        "extra-words"
      ],
      "reference_matches": [
        {
          "license_expression": "bsd-new",
          "license_expression_spdx": "BSD-3-Clause",
          "from_file": "diff-4.0.2.tgz/package/LICENSE",
          "start_line": 1,
          "end_line": 31,
          "matcher": "3-seq",
          "score": 93.89,
          "matched_length": 215,
          "match_coverage": 100.0,
          "rule_relevance": 100,
          "rule_identifier": "bsd-new_578.RULE",
          "rule_url": "https://github.com/nexB/scancode-toolkit/tree/develop/src/licensedcode/data/rules/bsd-new_578.RULE",
          "matched_text": "Software License Agreement (BSD License)\n\nCopyright (c) 2009-2015, Kevin Decker <kpdecker@gmail.com>\n\nAll rights reserved.\n\nRedistribution and use of this software in source and binary forms, with or without modification,\nare permitted provided that the following conditions are met:\n\n* Redistributions of source code must retain the above\n  copyright notice, this list of conditions and the\n  following disclaimer.\n\n* Redistributions in binary form must reproduce the above\n  copyright notice, this list of conditions and the\n  following disclaimer in the documentation and/or other\n  materials provided with the distribution.\n\n* Neither the name of Kevin Decker nor the names of its\n  contributors may be used to endorse or promote products\n  derived from this software without specific prior\n  written permission.\n\nTHIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS \"AS IS\" AND ANY EXPRESS OR\nIMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND\nFITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR\nCONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL\nDAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,\nDATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER\nIN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT\nOF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.",
          "matched_text_diagnostics": "Software License Agreement (BSD License)\n\n[Copyright] ([c]) [2009]-[2015], [Kevin] [Decker] <[kpdecker]@[gmail].[com]>\n\n[All] [rights] [reserved].\n\nRedistribution and use [of] [this] [software] in source and binary forms, with or without modification,\nare permitted provided that the following conditions are met:\n\n* Redistributions of source code must retain the above\n  copyright notice, this list of conditions and the\n  following disclaimer.\n\n* Redistributions in binary form must reproduce the above\n  copyright notice, this list of conditions and the\n  following disclaimer in the documentation and/or other\n  materials provided with the distribution.\n\n* Neither the name of [Kevin] [Decker] nor the names of its\n  contributors may be used to endorse or promote products\n  derived from this software without specific prior\n  written permission.\n\nTHIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS \"AS IS\" AND ANY EXPRESS OR\nIMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND\nFITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR\nCONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL\nDAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,\nDATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER\nIN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT\nOF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE."
        }
      ]
    }
    ```

**Comment by alok1304** [comment]
These failing test cases are passing on my system see
```
(venv) PS C:\Users\lenovo\Documents\GitHub\web dev\scancode-toolkit> pytest tests/scancode/test_cli.py                                
======================================================== test session starts =========================================================
platform win32 -- Python 3.11.0, pytest-7.1.2, pluggy-1.0.0
rootdir: C:\Users\lenovo\Documents\GitHub\web dev\scancode-toolkit, configfile: pyproject.toml
plugins: forked-1.4.0, xdist-2.5.0
collecting ...
74 tests selected, 5 tests skipped.
collected 79 items / 5 deselected / 74 selected

tests\scancode\test_cli.py ............................s....s.....ss.................................                           [100%]

====================================== 70 passed, 4 skipped, 5 deselected in 194.73s (0:03:14) ======================================= 
```

File: tests/scancode/data/extra-words/extra_words.expected.json [hunk 3]
```diff
@@ -0,0 +1,1010 @@
...
+            }
+          ],
+          "detection_log": [
+            "extra-words"
```
Comment (AyanSinhaMahapatra) [inline tests/scancode/data/extra-words/extra_words.expected.json]: @alok1304 we don't need to add the whole package archive as test, just having this LICENSE file is enough. We need to make sure tests are small and the minimum amount required to test the functionality.

File: tests/scancode/test_cli.py [hunk 4]
```diff
@@ -224,6 +224,16 @@ def test_scan_email_url_info():
     check_json_scan(expected, result_file, regen=REGEN_TEST_FIXTURES)
 
 
+def test_extra_words():
```
Comment (AyanSinhaMahapatra) [inline tests/scancode/test_cli.py]: Could you move this test to `tests/licensedcode`? Every module has it's own tests and functionality should be tested in correct location. Moving this to https://github.com/aboutcode-org/scancode-toolkit/blob/develop/tests/licensedcode/test_plugin_license_detection.py#L96 is appropriate, test data can also be moved to `tests/licensedcode/data/plugin_license/extra_words/`

File: src/licensedcode/detection.py [hunk 5]
```diff
@@ -1726,7 +1734,7 @@ def analyze_detection(license_matches, package_license=False):
     ):
         return DetectionCategory.LICENSE_CLUES.value
 
-    # Case where all matches have `matcher` as `1-hash` or `4-spdx-id`
+    # Case where all matches have `matcher` as `1-hash` or `4-spdx-id` or 2-aho
```
Comment (AyanSinhaMahapatra) [inline src/licensedcode/detection.py]: This is not correct, could you revert `is_correct_detection` back to:
``all(matcher in ("1-hash", "1-spdx-id") for matcher in matchers)``

The `2-aho` cases are meant to be caught below here, only if all other cases in between are not present:
```
    # Cases where Match Coverage is a perfect 100 for all matches
    else:
          return DetectionCategory.PERFECT_DETECTION.value
```
The only bug we needed to fix was in `get_detected_license_expression`, where we were missing catching the `analysis == DetectionCategory.EXTRA_WORDS.value` and thus the detection log not being populated..

File: tests/scancode/test_cli.py [hunk 6]
```diff
@@ -224,6 +224,16 @@ def test_scan_email_url_info():
...
+    test_dir = test_env.extract_test_tar('extra-words/diff-4.0.2.tgz')
+    result_file = test_env.get_temp_file('json')
+    args = ['--license','--license-text', '--license-text-diagnostics', '--license-diagnostics',
+        '--license-references','--todo', test_dir, '--json', result_file]
+    run_scan_click(args)
+    expected = test_env.get_test_loc('extra-words/extra_words.expected.json')
+    check_json_scan(expected, result_file, regen=REGEN_TEST_FIXTURES)    
+
```
Comment (AyanSinhaMahapatra) [inline tests/scancode/test_cli.py]: COuld you also add another test with the original example at: https://github.com/aboutcode-org/scancode-toolkit/issues/4400#issue-3091204633 from filesize? We want both `2-aho` and `3-seq` extra-words tested.

**Review by AyanSinhaMahapatra — CHANGES_REQUESTED** [review]
Thanks @alok1304, looking good mostly, ready to merge with a couple changes, see comments for more details.

Don't worry about the extra test failures, there are there because of https://github.com/aboutcode-org/scancode-toolkit/issues/4369, in the tests where we perform a force upgrade of all our dependencies and run the tests there. See https://github.com/aboutcode-org/scancode-toolkit/blob/develop/azure-pipelines.yml#L224

File: tests/scancode/data/extra-words/extra_words.expected.json (hunk 3 above)
Comment (alok1304) [reply tests/scancode/data/extra-words/extra_words.expected.json]: Okay got!!

File: tests/scancode/test_cli.py (hunk 4 above)
Comment (alok1304) [reply tests/scancode/test_cli.py]: thanks

**Comment by alok1304** [comment]
These failing test cases happening when i change this `all(matcher in ("1-hash", "1-spdx-id", "2-aho") for matcher in matchers)` to `all(matcher in ("1-hash", "1-spdx-id") for matcher in matchers)` in `is_correct_detection`.

see: https://github.com/aboutcode-org/scancode-toolkit/pull/4402/files#diff-5f6ccc1d8f9540cae715bddaf337246f4c69c48ca596b2086e05b18c541aef4aL1074

File: src/licensedcode/detection.py [hunk 7]
```diff
@@ -1071,7 +1072,7 @@ def is_correct_detection(license_matches):
     ]
 
     return (
-        all(matcher in ("1-hash", "1-spdx-id", "2-aho") for matcher in matchers)
```
Comment (AyanSinhaMahapatra) [inline src/licensedcode/detection.py]: > These failing test cases happening when i change this all(matcher in ("1-hash", "1-spdx-id", "2-aho") for matcher in matchers) to all(matcher in ("1-hash", "1-spdx-id") for matcher in matchers) in is_correct_detection.

Oops, my bad, this is actually effecting other parts and not correct. I thought you made the change to add `2-aho` here, but that was not correct, this was present from much earlier. Please revert back to your earlier changes and if tests are all green then this is ready to merge.

**Review by AyanSinhaMahapatra — CHANGES_REQUESTED** [review]
Thanks, please see comment below.

**Comment by alok1304** [comment]
@AyanSinhaMahapatra I did all modifications or changes that you said, in case of `2-aho` matcher, we didn't get `extra-words` in `detection_log`, otherwise all good.
Thanks for the review :)

Also, I squash all my commits into a single commit.

**Comment by AyanSinhaMahapatra** [comment]
@alok1304 could you add the `2-aho` test back? Also we need to cover both cases here for `detection_log` and see that it does not affect other tested cases.

**Comment by alok1304** [comment]
Hii @AyanSinhaMahapatra I do this things,
[`is_correct_detection`](https://github.com/aboutcode-org/scancode-toolkit/blob/develop/src/licensedcode/detection.py#L1062) This function is used in many places, so I write a new function i.e, is `is_correct_detection_2` This includes 
only  `1-hash` or `1-spdx-id`. 
Now, we got `extra-words` for both `2-aho` and `3-seq`. Also, all test cases are passed.

**Comment by alok1304** [comment]
@AyanSinhaMahapatra The file that you are providing there is the case in which we don't receive `extra-words`, Like for
```
/*
 * Copyright 2009-2011 Mozilla Foundation and contributors
 * Licensed under the New BSD license. See LICENSE.txt or:
 * http://opensource.org/licenses/BSD-3-Clause
 */
 ```
 This text present one of the file in https://registry.npmjs.org/source-map/-/source-map-0.6.1.tgz, 
 After scan, we got this output:
 ```
      "license_detections": [
        {
          "license_expression": "bsd-new",
          "license_expression_spdx": "BSD-3-Clause",
          "matches": [
            {
              "license_expression": "bsd-new",
              "license_expression_spdx": "BSD-3-Clause",
              "from_file": "package/source-map.js",
              "start_line": 3,
              "end_line": 4,
              "matcher": "3-seq",
              "score": 94.12,
              "matched_length": 16,
              "match_coverage": 100.0,
              "rule_relevance": 100,
              "rule_identifier": "bsd-new_359.RULE",
              "rule_url": "https://github.com/nexB/scancode-toolkit/tree/develop/src/licensedcode/data/rules/bsd-new_359.RULE",
              "matched_text": " * Licensed under the New BSD license. See LICENSE.txt or:\n * http://opensource.org/licenses/BSD-3-Clause",
              "matched_text_diagnostics": "Licensed under the New BSD license. See LICENSE.[txt] or:\n * http://opensource.org/licenses/BSD-3-Clause"
            }
          ],
          "detection_log": [],
          "identifier": "bsd_new-98c2f1ab-9c02-706d-1d86-fb586dfbfa57"
        }
      ]
 ```
 
 Here, due to `referenced_filenames` present in the rule that this executes first https://github.com/aboutcode-org/scancode-toolkit/blob/develop/src/licensedcode/detection.py#L1713
 
 I think we have to modify this.

**Review by AyanSinhaMahapatra — APPROVED** [review]
@alok1304 thanks++ LGTM!

This is fixed finally, and test failures are related to https://github.com/aboutcode-org/scancode-toolkit/issues/4369 which would be handled seperately, so merging!

**Comment by AyanSinhaMahapatra** [comment]
> Here, due to referenced_filenames present in the rule that this executes first https://github.com/aboutcode-org/scancode-toolkit/blob/develop/src/licensedcode/detection.py#L1713

> I think we have to modify this.

This is a seperate case, which is correctly detected as a `reference-to-other-file` which takes precedence over `extra-words`.
Here the best resolution is to add a rule with the correct text and referenced filename. I've added a rule for this at https://github.com/aboutcode-org/scancode-toolkit/pull/4405

We cannot use `extra-words` for referenced filename extensions, as these might then resolve to any file with the same name, which might not be what was being referenced.
